vllm/spec_decode/top1_proposer.py

```diff
@@ -226,47 +226,31 @@ def _merge_outputs(
         if maybe_sampler_output is None:
             # If no speculative tokens, the sampler output will be None.
             # In this case we return empty proposals.
-            proposal_tokens = torch.tensor(-1,
-                                           dtype=torch.long,
-                                           device=self._device).expand(
-                                               batch_size, proposal_len)
-            proposal_probs = torch.tensor(0,
-                                          dtype=torch.float32,
-                                          device=self._device).expand(
-                                              batch_size, proposal_len,
-                                              self._vocab_size)
-            proposal_lens_tensor = torch.tensor(0,
-                                                dtype=torch.long,
-                                                device=self._device).expand(
-                                                    len(proposal_lens))
+            proposal_tokens = torch.full((batch_size, proposal_len), -1,
+                                         dtype=torch.long, device=self._device)
+            proposal_probs = torch.zeros((batch_size, proposal_len, self._vocab_size),
+                                         dtype=torch.float32, device=self._device)
+            proposal_lens_tensor = torch.zeros(len(proposal_lens),
+                                               dtype=torch.long, device=self._device)
             return proposal_tokens, proposal_probs, proposal_lens_tensor
 
         sampler_output = maybe_sampler_output
-        proposal_tokens, proposal_probs, *_ = sampler_output_to_torch(
-            sampler_output, sampler_transposed)
-
-        # Now, reformat the output GPU tensors such that each sequence has
-        # a proposal. the proposal can be empty, e.g. [-1, -1, -1]
-
-        entire_proposal_tokens = proposal_tokens.new_full(
-            size=(batch_size, *proposal_tokens.shape[1:]),
-            fill_value=-1,
-        )
-        entire_proposal_tokens[nonzero_proposal_len_indices] = proposal_tokens
-        entire_proposal_probs = proposal_probs.new_zeros(
-            batch_size,
-            *proposal_probs.shape[1:],
-        )
-        entire_proposal_probs[nonzero_proposal_len_indices] = proposal_probs
-
-        proposal_tokens, proposal_probs = (
-            entire_proposal_tokens,
-            entire_proposal_probs,
-        )
+        proposal_tokens, proposal_probs, *_ = sampler_output_to_torch(sampler_output, sampler_transposed)
+
+        # Initialize tensors with appropriate sizes
+        entire_proposal_tokens = torch.full_like(proposal_tokens, -1, dtype=torch.long, device=self._device)
+        entire_proposal_probs = torch.zeros_like(proposal_probs, dtype=torch.float32, device=self._device)
```

Comment:
Is the first dim of these tensors always already `batch_size`? If so then this simplification makes sense!

Reply:
After checking several cases of speculative decoding, I found that when ngram was used, the first dim was not the batch size, so I restored the code for that part.